### PR TITLE
Fixes

### PR DESCRIPTION
**Pull request by kattrali**
Reduces complexity of locking/synchronization, which reduces scope of ANRs

### CODE REVIEW

File: sdk/src/main/java/com/bugsnag/android/AppData.java [hunk 1]
```diff
@@ -43,8 +43,8 @@ public void toStream(@NonNull JsonStream writer) throws IOException {
 
         writer.name("id").value(packageName);
         writer.name("buildUUID").value(config.getBuildUUID());
-        writer.name("duration").value(getDuration());
-        writer.name("durationInForeground").value(sessionTracker.getDurationInForeground(System.currentTimeMillis()));
+        writer.name("duration").value(getDurationMs()/1000);
+        writer.name("durationInForeground").value(sessionTracker.getDurationInForegroundMs(System.currentTimeMillis())/1000);
```
Comment (kattrali) [inline sdk/src/main/java/com/bugsnag/android/AppData.java]: This is not introduced in this changeset, but these values are different if the session is delivered immediately vs retried after a long attempted network request.

File: sdk/src/main/java/com/bugsnag/android/SessionTracker.java [hunk 2]
```diff
@@ -252,34 +243,41 @@ private void leaveBreadcrumb(String activityName, String lifecycleCallback) {
...
+     * @param nowMs The current time in ms
      */
-    void updateForegroundTracker(String activityName, boolean inForeground, long now) {
-        if (inForeground) {
-            long delta = now - lastForegroundMs;
+    void updateForegroundTracker(String activityName, boolean activityStarting, long nowMs) {
+        if (activityStarting) {
+            long noActivityRunningForMs = nowMs - activityLastStoppedAtMs.get();
+
+            //TODO:SM Race condition between isEmpty and put
```
Comment (kattrali) [inline sdk/src/main/java/com/bugsnag/android/SessionTracker.java]: I didn't find a real-world case for this, even programmatically switching foreground activities rapidly.

File: sdk/src/main/java/com/bugsnag/android/ErrorStore.java [hunk 3]
```diff
@@ -48,7 +48,7 @@ public int compare(File lhs, File rhs) {
     void flushOnLaunch(final ErrorReportApiClient errorReportApiClient) {
         final List<File> crashReports = findLaunchCrashReports();
 
-        if (crashReports.isEmpty() && config.getLaunchCrashThresholdMs() > 0) {
+        if (crashReports.isEmpty() && config.getLaunchCrashThresholdMs() == 0) {
             flushAsync(errorReportApiClient); // if disabled or no startup crash, flush async
```
[the remark this answers is not in the document]
Comment (fractalwrench) [reply sdk/src/main/java/com/bugsnag/android/ErrorStore.java]: This makes sense to me.

File: sdk/src/main/java/com/bugsnag/android/SessionTracker.java [hunk 4]
```diff
@@ -69,113 +65,116 @@
...
+        boolean notifyForRelease = configuration.shouldNotifyForReleaseStage(getReleaseStage());
+
+        if (notifyForRelease &&
+            (configuration.shouldAutoCaptureSessions() || !session.isAutoCaptured()) &&
+            session.isTracked().compareAndSet(false, true)) {
+            try {
+                Async.run(new Runnable() {
+                    @Override
+                    public void run() {
+                    //TODO:SM It would be good to optimise this
```
Comment (loopj) [inline sdk/src/main/java/com/bugsnag/android/SessionTracker.java]: indentation

Comment (loopj) [inline sdk/src/main/java/com/bugsnag/android/SessionTracker.java]: remove todo

File: sdk/src/main/java/com/bugsnag/android/SessionTracker.java [hunk 5]
```diff
@@ -69,113 +65,116 @@
...
-                Logger.warn("Invalid session tracking payload", e);
-                deleteStoredFiles(storedFiles);
+                List<File> storedFiles;
+
+                storedFiles = sessionStore.findStoredFiles();
+
+                if (!storedFiles.isEmpty()) {
+                    SessionTrackingPayload payload = new SessionTrackingPayload(storedFiles, client.appData);
+
+                    //TODO:SM Reduce duplication here and above
```
Comment (loopj) [inline sdk/src/main/java/com/bugsnag/android/SessionTracker.java]: remove todo

**Pull request merged**